pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,353 @@
+		// Get-WinEvent case as the command returns an error if a service is not logging to the Application provider.
+		if checkForNativeLogger(output, service) {
+			nativeLoggers = append(nativeLoggers, service)
+		} else {
+			fileLoggers = append(fileLoggers, service)
+		}
+	}
+	return nativeLoggers, fileLoggers
+}
+
+// copyServiceLogs invokes journalctl or Get-WinEvent with the provided args
+func (a *journalArgs) copyServiceLogs(ctx context.Context, w io.Writer, services []string, previousBoot int) {
+	cmdStr, args := getLoggingCmd(a, services, previousBoot)
+	cmd := exec.CommandContext(ctx, cmdStr, args...)
+	cmd.Stdout = w
+	cmd.Stderr = w
+
+	if err := cmd.Run(); err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			return
+		}
+		if previousBoot == 0 {
+			fmt.Fprintf(w, "error: journal output not available\n")
+		}
+	}
+}
+
+// copyFileLogs loops over all the services and attempts to collect the file logs of each service
+func copyFileLogs(ctx context.Context, w io.Writer, services []string) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	for _, service := range services {
+		if len(service) > 0 {
+			heuristicsCat(ctx, w, service)
+		}
+	}
+}
+
+// heuristicsCat attempts to collect logs from either
+// /var/log/service.log or
+// /var/log/service/service.log or
+// /var/log/service*INFO.* or
+// /var/log/service*INFO or
+// /var/log/service/service*INFO.* or
+// /var/log/service/service*INFO
+// in that order stopping on first success.
+func heuristicsCat(ctx context.Context, w io.Writer, service string) {
+	logDir := "/var/log/"
+	logFiles := [6]string{
+		path.Join(logDir, fmt.Sprintf("%s.log", service)),
+		path.Join(logDir, service, fmt.Sprintf("%s.log", service)),
+		path.Join(logDir, fmt.Sprintf("%s*INFO.*", service)),
+		path.Join(logDir, fmt.Sprintf("%s*INFO", service)),
+		path.Join(logDir, service, fmt.Sprintf("%s*INFO.*", service)),
+		path.Join(logDir, service, fmt.Sprintf("%s*INFO", service)),
+	}
+
+	for _, logFile := range logFiles {
+		if err := cat(ctx, w, logFile); err == nil {
+			break
+		}
+	}
+}
+
+// cat returns the contents of the given logFile
+func cat(ctx context.Context, w io.Writer, logFile string) error {
+	cmdStr, args := getCatCmd(logFile)
+	cmd := exec.CommandContext(ctx, cmdStr, args...)
+	cmd.Stdout = w
+	cmd.Stderr = w
+
+	return cmd.Run()
+}
+
+// parseQueryForServices traverses the query slice and returns the list of services
+func parseQueryForServices(query []string) (services []string) {
+	for _, q := range query {
+		if strings.ContainsAny(q, "/\\") {
+			continue
+		} else {
+			services = append(services, q)
+		}
+	}
+	return
+}
+
+func validateNodeLogQueryCombination(query url.Values) field.ErrorList {
+	queryingFiles, allErrs := validateQuery(query["query"])
+	if queryingFiles && (len(query.Get("sinceTime")) > 0 || len(query.Get("untilTime")) > 0 ||
+		len(query.Get("boot")) > 0 || len(query.Get("tailLines")) > 0 || len(query.Get("pattern")) > 0) {
```

Comment:
this is pretty difficult to read and would be easier if the url.Values → Options transformation was done first, and then we could just compare to an empty `Options{}` struct to see if any options were specified

Reply:
Addressed as part of the rework.